python/ql/src/experimental/semmle/python/frameworks/Django.qll

```diff
@@ -7,4 +7,260 @@ private import experimental.dataflow.DataFlow
 private import experimental.dataflow.RemoteFlowSources
 private import experimental.semmle.python.Concepts
 
-private module Django { }
+/**
+ * Provides models for the `django` PyPI package.
+ * See https://django.palletsprojects.com/en/1.1.x/.
+ */
+private module Django {
+  /** Gets a reference to the `django` module. */
+  private DataFlow::Node django(DataFlow::TypeTracker t) {
+    t.start() and
+    result = DataFlow::importNode("django")
+    or
+    exists(DataFlow::TypeTracker t2 | result = django(t2).track(t2, t))
+  }
+
+  /** Gets a reference to the `django` module. */
+  DataFlow::Node django() { result = django(DataFlow::TypeTracker::end()) }
+
+  /** Provides models for the `django` module. */
+  module django {
+    /** Gets a reference to the `django.db` module. */
+    private DataFlow::Node db(DataFlow::TypeTracker t) {
+      t.start() and
+      result = DataFlow::importNode("django.db")
+      or
+      t.startInAttr("db") and
+      result = django()
+      or
+      exists(DataFlow::TypeTracker t2 | result = db(t2).track(t2, t))
+    }
+
+    /** Gets a reference to the `django.db` module. */
+    DataFlow::Node db() { result = db(DataFlow::TypeTracker::end()) }
+
+    /** Provides models for the `django.db` module. */
+    module db {
+      /** Gets a reference to the `django.db.connection` object. */
+      private DataFlow::Node connection(DataFlow::TypeTracker t) {
+        t.start() and
+        result = DataFlow::importNode("django.db.connection")
+        or
+        t.startInAttr("connection") and
+        result = db()
+        or
+        exists(DataFlow::TypeTracker t2 | result = connection(t2).track(t2, t))
+      }
+
+      /** Gets a reference to the `django.db.connection` object. */
+      DataFlow::Node connection() { result = connection(DataFlow::TypeTracker::end()) }
+
+      /** Gets a reference to the `django.db.connection.cursor` class. */
+      private DataFlow::Node classCursor(DataFlow::TypeTracker t) {
+        t.start() and
+        result = DataFlow::importNode("django.db.connection.cursor")
+        or
+        t.startInAttr("cursor") and
+        result = connection()
+        or
+        exists(DataFlow::TypeTracker t2 | result = classCursor(t2).track(t2, t))
+      }
+
+      /** Gets a reference to the `django.db.connection.cursor` class. */
+      DataFlow::Node classCursor() { result = classCursor(DataFlow::TypeTracker::end()) }
+
+      /** Gets a reference to an instance of `django.db.connection.cursor`. */
+      private DataFlow::Node cursor(DataFlow::TypeTracker t) {
+        t.start() and
+        result.asCfgNode().(CallNode).getFunction() = classCursor().asCfgNode()
+        or
+        exists(DataFlow::TypeTracker t2 | result = cursor(t2).track(t2, t))
+      }
+
+      /** Gets a reference to an instance of `django.db.connection.cursor`. */
+      DataFlow::Node cursor() { result = cursor(DataFlow::TypeTracker::end()) }
```

Comment:
At this point, I think we're getting into dangerous waters in regards to naming, since there is no `django.db.connection.cursor` class. -- `connection` is defined in [`django/db/__init__.py`](https://github.com/django/django/blob/197b55c53469cf8344d1ba35175236780cb83bd1/django/db/__init__.py#L41), and does some funky forwarding.

I also don't think we've had to track the result of a function class like this before, but hopefully we can get to a nice solution :pray: 

Even though we can't even be sure that the `cursor` attribute is a class (it might just be a proxy function), I think modeling it as though it _is_ a class is very useful. However, I would would suggest we go with naming in the QLDoc along the lines of

```ql
/** Gets a reference to the `cursor` method on `django.db.connection`. */
DataFlow::Node classCursor() { result = classCursor(DataFlow::TypeTracker::end()) }


/** Gets a reference to a result of calling the `cursor` method on `django.db.connection`. */
DataFlow::Node cursor() { result = cursor(DataFlow::TypeTracker::end()) }
```

looking at the code, it's very strange that the `cursor()` predicate is the result of calling the `cursor` method... but I think if we adopt the `cursor::classRef` and `cursor::instance` pattern, it will become more obvious... maybe we should call this `cursor::methodRef` and `cursor::methodResult` instead?

Reply:
I see, `connection`, which should perhaps be supplemented by `connections`, ends up giving you a `ConnectionHandler` and `cursor` is just some property on that.

Reply:
Also adding support for connections would be great. (found some docs for it [here](https://docs.djangoproject.com/en/3.1/topics/db/multi-db/#using-raw-cursors-with-multiple-databases))